Add "Set as favourite" option for gridconfigs

matthiashamacher opened the pull request:
<!--

Before working on a contribution, you must determine on which branch you need to work:
- Bug fix: choose the latest maintenance branch, e.g. `10.0`
- Feature/Improvement: choose `10.x` 

> All bug fixes merged into the latest maintenance branch are also merged to the current dev branch (`10.x`) on a regular basis.

## Please make sure your PR complies with all of the following points: 
- [ ] Read and accept our [contributing guidelines](/CONTRIBUTING.md) before you submit a PR.
- [ ] Features need to be proper documented in `doc/` 
- [ ] Bugfixes need a short guide how to reproduce them -> target branch is the oldest supported maintenance branch, e.g. `10.0` (see Readme.md for the list of supported versions)
- [ ] Meet all coding standards (see PhpStan actions) 

**Don't submit a PR if it doesn't comply, it'll be closed without a comment!**
-->  
  

## Changes in this pull request  
Resolves [#9629](https://github.com/pimcore/pimcore/issues/9629)

## Additional info  
When the option is checked the grid will be set as the favourite for all shared users if they don't already have a favourite grid for that object. If they don't have a favourite grid at all the shared grid will also be set as the global favourite.

<img width="938" alt="CleanShot 2021-10-16 at 12 13 24@2x" src="https://user-images.githubusercontent.com/32356549/137583586-5b5c67af-870c-4233-a79b-ee6c8661db72.png">

Commit:
Add "Set as favourite" option for gridconfigs

Commit:
Code style, typos and finalize asset logic

mcop1 commented:
@matthiashamacher The "share globally" checkbox isn´t considered in the code. It only works if a user or role was selected from the corresponding dropdowns. Maybe we should consider this checkbox too? Or at least show a warning that this feature only works when a user/role was selected? What do you think?

matthiashamacher commented:
Oh yeah, you're right.
I think we should consider that checkbox too as otherwise, you need to add all users manually. This feature is for admins who want to set some defaults for the users, so in my opinion, this should work too.
But maybe we should implement that with kind of weight. What I mean is that currently if the user has already a favourite it is not overridden. But I think it would be useful to allow overriding if the current favourite of the user is a shared grid of another user (not owned by the current user) and is shared globally the favourite should be overridden.
I will definitely implement the logic for the "share globally" checkbox, what do you think of my proposal regarding the overriding?

mcop1 commented:
Sounds good, maybe an additional checkbox to activate/confirm the overwriting would be a good idea?

Commit:
Merge branch '10.x' into feature/set-grid-as-favorite

Commit:
Merge branch '10.x' into feature/set-grid-as-favorite

Commit:
Implement "Share globally" logic

matthiashamacher commented:
@mcop1 I've updated the PR. Please review again.

I've implemented the logic as followed:
If "share globally" and "set as favourite" are checked all users and roles are loaded and the grid config is set for all as favourite. If "share globally" is not checked the grid config is only set as favourite for the specified users and roles. If one of these users or roles already has a grid config it is checked 1. if the user is the owner. If that is the case we do not update the favourite. And 2. if the grid config is shared globally. If that is **not** the case we also do not update.

matthiashamacher commented on bundles/AdminBundle/Controller/Admin/DataObject/DataObjectHelperController.php:
```diff
@@ -991,6 +1000,119 @@ protected function updateGridConfigShares($gridConfig, $metadata)
+            // If the user has already a favourite for that object we check the current favourite and decide if we update
+            if ($favourite instanceof GridConfigFavourite) {
+                // Check if the user is the owner. If that is the case we do not update the favourite
+                if ((int) $favourite->getOwnerId() === $currentUser->getId()) {
```
We need to check against `$id` instead of `$currentUser->getId()`

matthiashamacher commented on bundles/AdminBundle/Controller/Admin/Asset/AssetHelperController.php:
```diff
@@ -614,6 +623,90 @@ protected function updateGridConfigShares($gridConfig, $metadata)
+
+            if ($favourite instanceof GridConfigFavourite) {
+                // Check if the user is the owner. If that is the case we do not update the favourite
+                if ((int) $favourite->getOwnerId() === $currentUser->getId()) {
```
We need to check against `$id` instead of `$currentUser->getId()`

Commit:
Fix override checks

mcop1 commented:
Looks very good now, just one more thing. Could you please confirm that "set as favourite" in combination with "shared roles" isn´t working? In my experience it gets shared, but never set as favourite.

Commit:
Remove role logic and clarify documentation

matthiashamacher commented:
Yes, you are right. I removed the role logic as it is not needed. Also, I added the information that it will only work for users to the documentation.

mcop1 commented:
Looks good 👍 thank you!
